torch/_inductor/codegen/cpp_utils.py

```diff
@@ -914,3 +914,62 @@ def _get_dtype_from_loopbodies(loop_bodies):
                     continue
                 dtypes.add(node.meta[OptimizationContext.key].dtype)
     return dtypes
+
+
+def _get_indexes_of_template_buf_read(epilogue_node, template_buf_name):
+    return [
+        read.index
+        for read in epilogue_node.get_reads()
+        if read.name == template_buf_name
+    ]
+
+
+def _check_supported_and_same_indexes(index_of_template_buf_read, epilogue_writes):
+    num_indexes = len(set(index_of_template_buf_read))
+
+    if num_indexes > 1:
+        same_index = False
+        supported = False  # Different read indexes not supported
+    elif num_indexes == 0:
+        same_index = True
+        supported = True  # No reads, automatically supported
+    elif num_indexes == 1:
+        index_of_template_buf_read = index_of_template_buf_read[0]
+        same_index = all(
+            write.index == index_of_template_buf_read for write in epilogue_writes
+        )
+        # TODO: Add support of fusion when the read of template buffer and the write of epilogue output
+        # in the epilogue node don't have the same index and change supported to True
+        supported = same_index
+    else:
+        raise AssertionError("Should not reach here")
+
+    return supported, same_index
+
+
+def _template_fusion_supported(template_node, epilogue_nodes):
+    assert isinstance(template_node, ir.CppTemplateBuffer)
+    template_buf_name = template_node.get_name()
+    indexes_of_template_buf_reads = [
+        _get_indexes_of_template_buf_read(epilogue_node, template_buf_name)
+        for epilogue_node in epilogue_nodes
+    ]
+    epilogue_nodes_writes = [
+        epilogue_node.get_read_writes().writes for epilogue_node in epilogue_nodes
+    ]
+
+    results = [
+        _check_supported_and_same_indexes(reads, writes)
+        for reads, writes in zip(indexes_of_template_buf_reads, epilogue_nodes_writes)
+    ]
+    supported, same_indexes = zip(*results)
+    return all(supported), all(same_indexes)
+
+
+def template_fusion_with_epilogues_supported(template, epilogues):
+    assert template.is_template()
+    template_node = template.get_template_node()
+    assert template_node is not None
+
+    epilogue_nodes = [n.node for epilogue in epilogues for n in epilogue.get_nodes()]
```

Comment:
I think we need to care about the case when some `epilogue_nodes` already fused into the `template`.